consensus/hotstuff/timeoutaggregator/timeout_collectors.go

```diff
@@ -0,0 +1,134 @@
+package timeoutaggregator
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/module/mempool"
+)
+
+// NewCollectorFactoryMethod is a factory method to generate a TimeoutCollector for concrete view
+type NewCollectorFactoryMethod = func(view uint64) (hotstuff.TimeoutCollector, error)
+
+// TimeoutCollectors implements management of multiple timeout collectors indexed by view.
+// Implements hotstuff.TimeoutCollectors interface. Creating a TimeoutCollector for a
+// particular view is lazy (instances are created on demand).
+// This structure is concurrently safe.
+// TODO: once VoteCollectors gets updated to stop managing worker pool we can merge VoteCollectors and TimeoutCollectors using generics
+type TimeoutCollectors struct {
+	log                zerolog.Logger
+	lock               sync.RWMutex
+	lowestRetainedView uint64                               // lowest view, for which we still retain a TimeoutCollector and process timeouts
+	collectors         map[uint64]hotstuff.TimeoutCollector // view -> TimeoutCollector
+	createCollector    NewCollectorFactoryMethod            // factory method for creating collectors
+}
+
+var _ hotstuff.TimeoutCollectors = (*TimeoutCollectors)(nil)
+
+func NewTimeoutCollectors(log zerolog.Logger, lowestRetainedView uint64, createCollector NewCollectorFactoryMethod) *TimeoutCollectors {
+	return &TimeoutCollectors{
+		log:                log,
+		lowestRetainedView: lowestRetainedView,
+		collectors:         make(map[uint64]hotstuff.TimeoutCollector),
+		createCollector:    createCollector,
+	}
+}
+
+// GetOrCreateCollector retrieves the hotstuff.TimeoutCollector for the specified
+// view or creates one if none exists.
+//  -  (collector, true, nil) if no collector can be found by the view, and a new collector was created.
+//  -  (collector, false, nil) if the collector can be found by the view
+//  -  (nil, false, error) if running into any exception creating the timeout collector state machine
+// Expected error returns during normal operations:
+//  * mempool.DecreasingPruningHeightError - in case view is lower than lowestRetainedView
```

Comment:
In this application, we're using `DecreasingPruningHeightError` for a slightly different purpose than its documentation and other uses.

The [Godoc](https://github.com/onflow/flow-go/blob/1f7ce61f6a6fbe1f0cbf708de2712ae416533442/module/mempool/errors.go#L37-L38) for `DecreasingPruningHeightError` says:
```go
// DecreasingPruningHeightError indicates that we are pruning a mempool by a height that
// is lower than existing height
```

But here (and in most usages of the error type), we return the error when we are querying, not when we are pruning. I'd suggest to change the error name and documentation to correspond better to the usage:
```go
// BelowPrunedThresholdError indicates that we are attempting to query or prune a mempool by a 
// key (typically block height or block view) which is lower than the lowest retained key threshold. 
// In other words, we have already pruned above the queried key.
type BelowPrunedThresholdError ...
```

Reply:
https://github.com/onflow/flow-go/pull/2740/commits/35d88f2fc769aa3d99586fba43d4830c6cf0c132